databricks_cli/pipelines/api.py

```diff
@@ -0,0 +1,187 @@
+                    raise RuntimeError('Invalid file uri scheme')
+                local_lib_objects.append(LibraryObject(lib_object.lib_type, lib_object.path[5:]))
+            else:
+                rest_lib_objects.append(lib_object)
+        return local_lib_objects, rest_lib_objects
+
+    @staticmethod
+    def _get_hashed_path(path):
+        """
+        Finds the corresponding dbfs file path for the file located at the supplied path by
+        calculating its hash.
+        :param path: Local File Path
+        :return: Remote Path (pipeline_base_dir + file_hash (dot) file_extension)
+        """
+        hash_buffer = sha1()
+        try:
+            with open(path, 'rb') as f:
+                while True:
+                    data = f.read(BUFFER_SIZE)
+                    if not data:
+                        break
+                    hash_buffer.update(data)
+        except Exception as e:
+            raise RuntimeError('Error \'{}\' while processing {}'.format(e, path))
+
+        file_hash = hash_buffer.hexdigest()
+        path = '{}{}{}'.format(base_pipelines_dir, file_hash, os.path.splitext(path)[1])
+        return path
+
+    def _get_files_to_upload(self, local_lib_objects, remote_lib_objects):
+        """
+        Returns a Local/Remote pair for every file that needs to be uploaded to dbfs. Files are
+        stored under base_pipelines_dir defined above, and we only upload files that don't already
+        exist in dbfs.
+        :param local_lib_objects: List[LibraryObject]
+        :param remote_lib_objects: List[LibraryObject]
+        :return: List[(LibraryObject, LibraryObject)]
+        """
+        transformed_remote_lib_objects = map(
+            lambda rlo: LibraryObject(rlo.lib_type, DbfsPath(rlo.path)),
+            remote_lib_objects)
+        return list(filter(lambda lo_tuple: not self.dbfs_client.file_exists(lo_tuple[1].path),
+                           zip(local_lib_objects, transformed_remote_lib_objects)))
+
+    @staticmethod
+    def _get_credentials_for_request():
+        """
+        Only required while the deploy/delete APIs require credentials in the body as well
+        as the header. Once the API requirement is relaxed, this function can be stripped out and
+        includes for this function removed.
+        """
+        profile = get_profile_from_context()
+        if profile:
+            config = ProfileConfigProvider.get_config(profile)
+        else:
+            config = get_config()
+        if not config or not config.is_valid:
+            raise InvalidConfigurationError.for_profile(profile)
+
+        if config.is_valid_with_token:
+            return {'token': config.token}
+        else:
+            return {'user': config.username, 'password': config.password}
+
+
+class LibraryObject(object):
+    def __init__(self, lib_type, lib_path):
+        self.path = lib_path
+        self.lib_type = lib_type
+
+    @classmethod
+    def convert_from_libraries(cls, libraries):
```

Comment:
nit: can we rename this to something a bit more descriptive? `from_libraries` is a bit weird since you're returning a list of LibraryObjects. Maybe `from_json`?

Reply:
This is specifically library json objects (not any json). I'm thinking of renaming it to from_json_libraries to be more clear. Does that work?

Reply:
I don't feel strongly about it but I think `from_json` is better still. Library is implicit from `LibraryObjects.from_json`. Your choice though.